libraries/Microsoft.Bot.Builder.Dialogs/Prompts/CloudOAuthPrompt.cs

```diff
@@ -0,0 +1,137 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Collections.Generic;
+using System.Globalization;
+using System.Threading;
+using System.Threading.Tasks;
+using Microsoft.Bot.Connector.Authentication;
+using Microsoft.Bot.Schema;
+
+namespace Microsoft.Bot.Builder.Dialogs
+{
+    /// <summary>
+    /// This is replacement for the OauthPrompt that replies on the UserTokenClient it finds in the TurnState.
+    /// </summary>
+    internal class CloudOAuthPrompt : Dialog
+    {
+        private readonly OAuthPromptSettings _settings;
+        private readonly PromptValidator<TokenResponse> _validator;
+
+        /// <summary>
+        /// Initializes a new instance of the <see cref="CloudOAuthPrompt"/> class.
+        /// </summary>
+        /// <param name="settings">Additional OAuth settings to use with this instance of the prompt.</param>
+        /// <param name="validator">A custom validator that can be used against Message activities.</param>
+        public CloudOAuthPrompt(OAuthPromptSettings settings, PromptValidator<TokenResponse> validator)
+        {
+            _settings = settings;
+            _validator = validator;
+        }
+
+        /// <summary>
+        /// Called when a prompt dialog is pushed onto the dialog stack and is being activated.
+        /// </summary>
+        /// <param name="dc">The dialog context for the current turn of the conversation.</param>
+        /// <param name="options">Optional, additional information to pass to the prompt being started.</param>
+        /// <param name="cancellationToken">A cancellation token that can be used by other objects or threads to receive notice of cancellation.</param>
+        /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+        /// <remarks>If the task is successful, the result indicates whether the prompt is still
+        /// active after the turn has been processed by the prompt.</remarks>
+        public override async Task<DialogTurnResult> BeginDialogAsync(DialogContext dc, object options = null, CancellationToken cancellationToken = default)
+        {
+            var userTokenClient = dc.Context.TurnState.Get<UserTokenClient>() ?? throw new InvalidOperationException("The UserTokenClient is not supported by the current adapter.");
+
+            var tokenResponse = await userTokenClient.GetUserTokenAsync(dc.Context.Activity.From.Id, _settings.ConnectionName, dc.Context.Activity.ChannelId, null, cancellationToken).ConfigureAwait(false);
+            if (tokenResponse != null)
```

Comment:
It would be nice as part of this re-write to get some comments explaining what happens at the high level. For xample, in this case, we could say:  `// We already have a token, no need to show a prompt`. We don't need, however, to fill everything with comments. I'd say the following rule: if you had to think for a few mins about what a piece of code means after reading the bare code, then a comment will really help future maintainers.